overlord/snapshotstate/backend/backend.go

```diff
@@ -486,34 +498,52 @@ var ErrCannotCancel = errors.New("cannot cancel: import already finished")
 // a commit.
 type importTransaction struct {
 	id        uint64
+	flockPath string
+	// flock is a file lock associated with the import, the logic
+	// using it prevents potentially cleaning up/cancelling a
+	// still in progress import in case we started allowing
+	// cleaning and imports to happen concurrently
+	flock     *osutil.FileLock
 	committed bool
 }
 
+// newImportTransaction creates a new importTransaction for the given
+// snapshot id.
 func newImportTransaction(setID uint64) *importTransaction {
-	return &importTransaction{id: setID}
+	return &importTransaction{
+		flockPath: filepath.Join(dirs.SnapshotsDir, fmt.Sprintf(importingFnFmt, setID)),
+		id:        setID,
+	}
 }
 
-func (t *importTransaction) importInProgressFilesGlob() string {
-	return filepath.Join(dirs.SnapshotsDir, fmt.Sprintf("%d_*.zip", t.id))
+// newImportTransactionFromImportFile creates a new importTransaction
+// for the given import file path. It may return an error if an
+// invalid file was specified.
+func newImportTransactionFromImportFile(p string) (*importTransaction, error) {
+	parts := importingFnRegexp.FindStringSubmatch(path.Base(p))
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("cannot determine snapshot id from %q", p)
+	}
+	setID, err := strconv.ParseUint(parts[1], 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	return newImportTransaction(setID), nil
 }
 
 // Start marks the start of a snapshot import
 func (t *importTransaction) Start() error {
-	return ioutil.WriteFile(importInProgressFilepath(t.id), nil, 0644)
+	return t.lock()
 }
 
 // InProgress returns true if there is an import for this transactions
 // snapshot ID already.
 func (t *importTransaction) InProgress() bool {
-	return osutil.FileExists(importInProgressFilepath(t.id))
-}
-
-func importInProgressFilepath(setID uint64) string {
-	return filepath.Join(dirs.SnapshotsDir, fmt.Sprintf("%d_importing", setID))
-}
-
-func importInProgressFor(setID uint64) bool {
-	return osutil.FileExists(importInProgressFilepath(setID))
+	if flock, err := osutil.OpenExistingLockForReading(t.flockPath); err == nil {
+		defer flock.Close()
+		return flock.TryLock() == osutil.ErrAlreadyLocked
+	}
+	return false
 }
 
 // Cancel cancels a snapshot import and cleanups any files on disk belonging
```

Comment:
I don't know if it's needed to comment on this but with the latest changes Cancel can only be called on an  in-progress import. Before I think it would have done nothing instead of erroring as now.